packages/pwa-kit-dev/src/ssr/server/build-dev-server.js

```diff
@@ -278,11 +278,15 @@ export const DevServerMixin = {
 
     render(req, res, next) {
         const app = req.app
-        if (app.__webpackReady()) {
-            app.__hotServerMiddleware(req, res, next)
-        } else {
-            this._redirectToLoadingScreen(req, res, next)
-        }
+
+        // Ensure that we do not try to render anything until the webpack bundle is valid
+        app.__devMiddleware.waitUntilValid(() => {
+            if (app.__webpackReady()) {
+                app.__hotServerMiddleware(req, res, next)
+            } else {
+                this._redirectToLoadingScreen(req, res, next)
+            }
+        })
```

Comment:
```suggestion
        app.__devMiddleware.waitUntilValid(() => {
            app.__hotServerMiddleware(req, res, next)
        })
```
`waitUntilValid` [checks if `context.state` is set](https://github.com/webpack/webpack-dev-middleware/blob/034797863fc414d7c30cfc2b883a2f082436e1ec/src/utils/ready.js#L13). `__webpackReady` does the [same check](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/79c15513e4c8065d8ad44027711159192ff8011d/packages/pwa-kit-dev/src/ssr/server/build-dev-server.js#L145), so the redirect will never happen.

Reply:
Also, if we get rid of `_redirectToLoadingScreen` here, we can delete the method, because it's not used anywhere else.

Reply:
It seems that we call `_redirectToLoadingScreen` directly for a test so I've left it in for now: https://github.com/SalesforceCommerceCloud/pwa-kit/blob/79c15513e4c8065d8ad44027711159192ff8011d/packages/pwa-kit-dev/src/ssr/server/build-dev-server.test.js#L139

Reply:
I think @wjhsf is probably right and that since we're removing the functionality this test is testing the thing that we're **explicitly not doing anymore** (e.g. can delete the test)

https://github.com/SalesforceCommerceCloud/pwa-kit/blob/79c15513e4c8065d8ad44027711159192ff8011d/packages/pwa-kit-dev/src/ssr/server/build-dev-server.test.js#L135C1-L148C3